### PR TITLE
Generalize Handler interface for 1D & 2D data

### PR DESCRIPTION
**Pull request by gibsramen**
Suggestion from @ElDeveloper that the way Evident works, any univariate data could be used instead of alpha diversity values (e.g. log ratios). As such, we can generalize the code (and especially the Q2 plugin) so that more data types can be used.

I propose renaming `AlphaDiversityHandler` and `BetaDiversityHandler` to `UnivariateDataHandler` and `BivariateDataHandler` respectively. The main examples will still use alpha and beta diversity but we can add a log-ratio example as well.

@ElDeveloper @wasade Do you think this renaming makes sense?

TODO:

- [x] Update Q2 plugin
- [x] Update README
- [x] Make sure interactive app still works
- [x] Update images

### CODE REVIEW

**Comment by ElDeveloper** [comment]
What do you think about `Multivariate...` instead of `Bivariate...`? Beta diversity in general is a multivariate data analysis approach so I think it makes sense.

**Comment by gibsramen** [comment]
I was thinking bivariate in the sense that the input data (distance matrix) has two dimensions. Multivariate might make more sense, though.

File: README.md [hunk 1]
```diff
@@ -193,16 +193,35 @@ If not, we recommend you read the excellent [documentation](https://docs.qiime2.
...
 
-As with the standalone version, Evident requires a diversity file and a sample metadata file.
-These inputs are expected to conform to QIIME 2 standards.
+### A note on univariate data
+
+Because of some design decisions in QIIME 2, we cannot use a generic `SampleData` input artifact for univariate analysis.
+However, we are also required to specify an input, even if it is optional.
+As a result, there are two ways to perform univariate data analysis in the QIIME 2 version of Evident (for both power analysis and effect size calculations).
+
+1. Provide a `SampleData[AlphaDiversity` artifact to `--i-data`.
```
Comment (ElDeveloper) [inline README.md]:
```suggestion
1. Provide a `SampleData[AlphaDiversity]` artifact to `--i-data`.
```

File: evident/q2/_methods.py [hunk 2]
```diff
@@ -1,101 +1,137 @@
...
+    else:
+        if data_column is None:
+            raise ValueError(
+                "If not providing an input for data, a value must be "
+                "provided for data_column to use a sample metadata column "
+                "identifier."
+            )
+        if data_column not in sample_metadata.columns:
+            raise ValueError(f"{data_column} not found in sample metadata.")
+        if sample_metadata[data_column].dtype.kind not in ["i", "f"]:
```
Comment (ElDeveloper) [inline evident/q2/_methods.py]: I think pandas' `select_dtypes` would do the trick here, check this SO answer:
https://stackoverflow.com/a/43982712/379593

File: evident/q2/plugin_setup.py [hunk 3]
```diff
@@ -150,14 +163,16 @@
...
 rm_param_descs["correlation"] = "Correlation between repeated measurements."
 rm_param_descs["epsilon"] = "Sphericity parameter."
+rm_param_descs["data_column"] = "Column in metadata containing data."
 
 plugin.methods.register_function(
-    function=alpha_power_analysis_repeated_measures,
-    inputs={"alpha_diversity": SampleData[AlphaDiversity]},
-    input_descriptions={"alpha_diversity": "Alpha diversity vector"},
+    function=univariate_power_analysis_repeated_measures,
+    inputs={"data": SampleData[AlphaDiversity]},
```
Comment (ElDeveloper) [inline evident/q2/plugin_setup.py]: I don't think you need a specific input for this. As far as I recall you should be able to transform AlphaDiversity on-the-fly into a "metadata column". Or am I missing something here?

**Review by ElDeveloper — CHANGES_REQUESTED** [review]
Looks great, just a few minor changes.

File: evident/q2/plugin_setup.py (hunk 3 above)
Comment (gibsramen) [reply evident/q2/plugin_setup.py]: As far as I can tell Q2 required an input semantic type, even if it is optional. It also seems like you can't provide Metadata, only a Q2 semantic type.

`TypeError: register_function() missing 1 required positional argument: 'inputs'`

Comment (ElDeveloper) [reply evident/q2/plugin_setup.py]: That's odd, based on what I see on `q2-metadata` [inputs are not required for actions](https://github.com/qiime2/q2-metadata/blob/master/q2_metadata/plugin_setup.py#L32). If I understand this right, the way to set this up for evident is to remove the inputs, and leave just the metadata parameters so the Q2 machinery can transform and join the artifacts on-the-fly.

Comment (gibsramen) [reply evident/q2/plugin_setup.py]: Gotcha, gotcha. What I was doing was just removing the `input` lines completely. I think if we keep those lines but just pass empty dictionaries it should work. I'll try it out and update. Thanks!!!

Comment (ElDeveloper) [reply evident/q2/plugin_setup.py]: Excellent!

**Comment by gibsramen** [comment]
Just noticed a bug with the interactive - gonna address that and then we should be good.

File: evident/q2/plugin_setup.py [hunk 4]
```diff
@@ -81,45 +80,52 @@
...
     ),
     package="evident"
 )
 
 
+UNIV_PA_PARAM_DESCS = PA_PARAM_DESCS.copy()
+UNIV_PA_PARAM_DESCS["data_column"] = "Column in metadata containing data."
+
+# QIIME 2 needs an input so we provide SampleData[AlphaDiversity] which is
+# optional. Can't provide just SampleData, unfortunately.
```
Comment (ElDeveloper) [inline evident/q2/plugin_setup.py]:
```suggestion
```

**Review by ElDeveloper — CHANGES_REQUESTED** [review]
Just one minor change, otherwise looks good to be merged.

File: evident/q2/plugin_setup.py (hunk 4 above)
Comment (gibsramen) [reply evident/q2/plugin_setup.py]: Good catch!

**Comment by gibsramen** [comment]
Thanks so much @ElDeveloper !!!

**Comment by ElDeveloper** [comment]
Thank you @gibsramen

**Pull request merged**